File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-java/src/main/java/com/opspilot/llm/ResilientChatModel.java
package com.opspilot.llm;

import com.opspilot.domain.Exceptions;
import com.opspilot.tools.Tool;

import java.util.List;

/**
 * Porte de OpsResilientChatModel / composeResilientRunnable (agents/model.ts):
 * retry no primario -> fallback pro backup (se configurado) -> lanca
 * ModelUnavailableException se ambos falharem apos as tentativas.
 */
public class ResilientChatModel implements ChatModel {

    public static final int RETRY_ATTEMPTS = 2;

    private final ChatModel primary;
    private final ChatModel backup;

    public ResilientChatModel(ChatModel primary, ChatModel backup) {
        this.primary = primary;
        this.backup = backup;
    }

    @Override
    public ModelResponse invoke(List<ChatMessage> messages, List<Tool> tools) {
        RuntimeException primaryError = null;
        for (int attempt = 0; attempt < RETRY_ATTEMPTS; attempt++) {
            try {
                return primary.invoke(messages, tools);
            } catch (RuntimeException e) {
                primaryError = e;
            }
        }

        if (backup == null) {
            throw new Exceptions.ModelUnavailableException(messageOf(primaryError));
        }

        RuntimeException backupError = null;
        for (int attempt = 0; attempt < RETRY_ATTEMPTS; attempt++) {
            try {
                return backup.invoke(messages, tools);
            } catch (RuntimeException e) {
                backupError = e;
            }
        }
        throw new Exceptions.ModelUnavailableException(messageOf(backupError));
    }

    private static String messageOf(RuntimeException error) {
        return error != null ? error.getMessage() : "All configured language models failed";
    }
}

File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-go/llm/resilient_chat_model.go
package llm

import (
	"opspilot/domain"
	"opspilot/tools"
)

// RetryAttempts é o porte de ResilientChatModel.RETRY_ATTEMPTS (porte Java).
const RetryAttempts = 2

// ResilientChatModel é o porte de OpsResilientChatModel /
// composeResilientRunnable (agents/model.ts): retry no primário -> fallback
// pro backup (se configurado) -> retorna ModelUnavailableError se ambos
// falharem após as tentativas. Backup nil equivale a "sem backup configurado"
// (null no original).
type ResilientChatModel struct {
	primary ChatModel
	backup  ChatModel
}

var _ ChatModel = (*ResilientChatModel)(nil)

func NewResilientChatModel(primary, backup ChatModel) *ResilientChatModel {
	return &ResilientChatModel{primary: primary, backup: backup}
}

func (r *ResilientChatModel) Invoke(messages []ChatMessage, availableTools []tools.Tool) (ModelResponse, error) {
	var lastErr error

	for attempt := 0; attempt < RetryAttempts; attempt++ {
		response, err := r.primary.Invoke(messages, availableTools)
		if err == nil {
			return response, nil
		}
		lastErr = err
	}

	if r.backup == nil {
		return ModelResponse{}, &domain.ModelUnavailableError{Message: messageOf(lastErr)}
	}

	for attempt := 0; attempt < RetryAttempts; attempt++ {
		response, err := r.backup.Invoke(messages, availableTools)
		if err == nil {
			return response, nil
		}
		lastErr = err
	}
	return ModelResponse{}, &domain.ModelUnavailableError{Message: messageOf(lastErr)}
}

func messageOf(err error) string {
	if err != nil {
		return err.Error()
	}
	return "All configured language models failed"
}
